src/media/media.ts

```diff
@@ -0,0 +1,820 @@
+   */
+  async startDisplayMedia (): Promise<MediaStream> {
+    const constraints = this.getScreenShareConstraints();
+    const promise = this.hasGetDisplayMedia()
+      ? window.navigator.mediaDevices.getDisplayMedia(constraints)
+      : window.navigator.mediaDevices.getUserMedia(constraints);
+
+    const stream = await promise.catch(e => {
+      /* we want to emit errors on `sdk.on('sdkError')` */
+      createAndEmitSdkError.call(this.sdk, SdkErrorTypes.media, e);
+      throw e;
+    });
+
+    stream.getTracks().forEach(t => this.allMediaTracksCreated.push(t));
+    return stream;
+  }
+
+  /**
+   * Create media with video and/or audio. See `interface IMediaRequestOptions`  
+   *  for more information about available options. 
+   * 
+   * It is _highly_ recommended that `sdk.media.requestPermissions(type)` be called
+   *  with each desired media type _before_ using `startMedia`. This will ensure 
+   *  all media permissions have been granted before starting media. 
+   * 
+   * If `retry` is `true` (which is default), the SDK will attempt to call 
+   *  `startMedia()` again on any failures. Two notes about this behavior: 
+   *    1) It will _not_ retry permissions errors
+   *    2) It will retry media with system defaults and no other options
+   *        (such as `deviceId`s, `videoFrameRate`, & `videoResolution`)
+   * 
+   * Warning: if `sdk.media.requestPermissions(type)` has NOT been called before 
+   *  calling `startMedia`, this function will call `sdk.media.requestPermissions(type)`.
+   *  If calling `startMedia` with both `audio` and `video` _before_ requesting permissions,
+   *  `startMedia` will attempt to gain permissions for `audio` first and then `video` (because
+   *  media permissions must be requested one at a time). If `audio` fails, it will 
+   *  not attempt to gain permissions for `video` – the error will stop execution. 
+   * 
+   * @param opts video and/or audio default device or deviceId. Defaults to `{video: true, audio: true}`
+   * @param retry whether the sdk should retry on an error
+   */
+  async startMedia (opts: IMediaRequestOptions = { video: true, audio: true }, retry: boolean = true): Promise<MediaStream> {
+    /* this will set media type to `truthy` if `null` was passed in */
+    const constraints: any = this.getStandardConstraints(opts);
+    const conversationId = opts.session?.conversationId;
+    const sessionId = opts.session?.id;
+    const isFirefox = browserama.isFirefox;
+    const requestingVideo = opts.video || opts.video === null;
+    const requestingAudio = opts.audio || opts.audio === null;
+
+    const { micPermissionsRequested, cameraPermissionsRequested } = this.getState();
+    let mediaStreamFromPermissions: MediaStream;
+
+    /* if we haven't requested permissions for audio yet, we need to do that first */
+    if (requestingAudio && !micPermissionsRequested) {
+      const optsWithOutsession = opts;
+      delete opts.session;
```

Comment:
I don't think this is doing what you expected. I would expect this to remove the session from `opts` and `optsWithOutSession` (also, i think it should be `optsWithoutSession`). I think you should copt the opts object. ex:
`const optsWithoutSession = { ...opts };`